### ISSUE TITLE
`matmul` (and others) Precision issues between Candle & PyTorch

### ISSUE THREAD

**Issue by andrew-shc** [huggingface/candle#3032]
We noticed there's some precision discrepancy in matrix multiplication and the linear layer between between Candle and PyTorch. This matters a lot when reproducing LLMs originated from PyTorch into Candle. We used the `hf_hub::api::Api` to get the safetensors from the hub and for testing the precision issues for each modules independently. This also occurs for the `BF16` dtype in `Cuda`.

Here's a shortened list of tests (for brevity) between `candle_core::tensor::Tensor::matmul` and `torch.matmul`
```
❌ test_0: MSE=0.0000000004096404, MAE=0.00001550 (dims: 2048x256, dtype: F32, device: Cpu)
❌ test_1: MSE=0.0000000003628351, MAE=0.00001453 (dims: 2048x256, dtype: F32, device: Cpu)
...
❌ test_48: MSE=0.0000000000824194, MAE=0.00000633 (dims: 512x1024, dtype: F32, device: Cpu)
❌ test_49: MSE=0.0000000003840639, MAE=0.00001534 (dims: 2048x256, dtype: F32, device: Cpu)
```

We did notice `candle_nn::Embedding` performed at 0-tolerance (tested indirectly), which probably means the the loaded weights themselves are working precisely.

Have you guys tried validating your implementation with the PyTorch at 0-tolerance (within the same CPU/GPU architecture)? Is there any proper way to mitigate this? We need it for our implementation. Thank you.

**Comment by robertknight** [comment]
> Have you guys tried validating your implementation with the PyTorch at 0-tolerance (within the same CPU/GPU architecture)? Is there any proper way to mitigate this? We need it for our implementation. 

By 0-tolerance do you mean that you are expecting Candle and PyTorch to produce _exactly_ (as in, bit-for-bit) the same results for f32 matrix multiplication? That's not a valid expectation given that floating point addition and multiplication are not associative. There will be very small differences depending on the order in which calculations are done. This is true even just across different PyTorch releases.

**Comment by andrew-shc** [comment]
Would there be a way to make sure the activations of Candle is same as activations of layers in PyTorch (the standard)? Or just at least the matmul?

For context, this is for LLMs, and we notice the errors accrue layer by layer for each token until one point it generates a different token, diverting the remaining output from the original standard PyTorch implementation. (We already made sure random sampling is turned off and double checked on token order for same logits)

> By 0-tolerance do you mean that you are expecting Candle and PyTorch to produce exactly (as in, bit-for-bit) the same results for f32 matrix multiplication? 

Yes (f32 & bf16), even if we need to compare with the same PyTorch release. We're only using it for inference so the reproducibility of quality is important.

**Comment by robertknight** [comment]
> Would there be a way to make sure the activations of Candle is same as activations of layers in PyTorch (the standard)? Or just at least the matmul?

Two different implementations of float matrix multiplication are unlikely to produce results that match bit-for-bit but they should be very close. You can see this if you compare PyTorch to numpy:

```py
import numpy as np
import torch

lhs = np.random.rand(512, 512).astype(np.float32)
rhs = np.random.rand(512, 512).astype(np.float32)
np_result = np.matmul(lhs, rhs)
torch_result = torch.matmul(torch.tensor(lhs), torch.tensor(rhs))
print("max difference (torch vs numpy): ", (torch_result - np_result).max())
```

For me at least, the difference is very small but non-zero.

It is possible there is an issue that is causing the differences between the implementations to be larger than expected. One way to check would be to compare with a third implementation or perform the computation in higher precision to get a reference result and see how much each implementation's output (in lower precision) differs.

**Comment by andrew-shc** [comment]
Ok thanks
